WooCommerce/Classes/Tools/InfiniteScroll/PaginationTracker.swift

```diff
@@ -0,0 +1,152 @@
+import Yosemite
+
+/// Delegate of `PaginationTracker` that implements syncing per page number and size.
+protocol PaginationTrackerDelegate: class {
+    typealias SyncCompletion = (Result<Bool, Error>) -> Void
+
+    /// Syncs a page by the number and size, and returns an error or whether there might be the next page on completion.
+    ///
+    /// - Parameter reason: A value passed from `resync` or `syncFirstPage`. This can
+    ///                     be used to decide how to perform the sync.
+    /// - Parameter onCompletion: called on sync completion that returns either an error or a boolean that indicates
+    ///                           whether there might be the next page to sync.
+    func sync(pageNumber: Int, pageSize: Int, reason: String?, onCompletion: SyncCompletion?)
+}
+
+/// Keeps track of the pagination for API syncing to support infinite scroll.
+final class PaginationTracker {
+    /// Default Settings
+    private enum Defaults {
+        static let pageFirstIndex = Store.Default.firstPageNumber
```

Comment:
Imagine all the bugs we can cause if we change the value of `Store.Default.firstPageNumber`. 🤣

Reply:
yeaa suddenly many screens wouldn't get the first page of data, and we will get so many reviews/reports about the outdated data 😅 😨